Lib/glyphsLib/builder.py

```diff
@@ -695,17 +695,44 @@ def add_groups_to_ufo(ufo, kerning_groups):
 
 def build_gdef(ufo):
     """Build a table GDEF statement for ligature carets."""
-    carets = {}
+    bases, ligatures, marks, components, carets = set(), set(), set(), set(), {}
     for glyph in ufo:
+        has_caret = False
         for anchor in glyph.anchors:
             name = anchor.get('name')
             if name and name.startswith('caret_') and 'x' in anchor:
+                has_caret = True
                 carets.setdefault(glyph.name, []).append(str(anchor['x']))
-    if not carets:
+        glyphinfo = glyphsLib.glyphdata.get_glyph(glyph.name)
+        if glyphinfo:
+            category, subCategory = glyphinfo.category, glyphinfo.subCategory
+        else:
+            category, subCategory = None, None
+        # TODO: Find out how Glyphs.app assigns ligature components.
+        # Currently, we leave components empty but that is probably wrong.
+        # https://github.com/googlei18n/glyphsLib/issues/99
+        if has_caret or subCategory == 'Ligature':
+            ligatures.add(glyph.name)
+        elif category == 'Mark' and (subCategory == 'Nonspacing' or
+                                     subCategory == 'Spacing Combining'):
+            # https://github.com/googlei18n/ufo2ft/issues/91#issuecomment-266530863
+            marks.add(glyph.name)
+        else:
+            bases.add(glyph.name)
```

Comment:
From what I understand this "Base" class is not interpreted as residual (i.e. anything non-Mark or non-Ligature). A user could define "Base" class selectively, as in the example from the GDEF spec where one only assigns class values for Arabic glyphs, but not for Latin -- which are implicitly assigned class 0, the non-class.
When writing these features manually, one does what is most suitable for the case at hand.
But the usual question for glyphsLib is, how does Glyphs.app generates this classes automatically?

For example, by reading Georg's comment in https://github.com/googlei18n/glyphsLib/issues/99#issuecomment-275087699, it seems like anchors' `name.startswith('caret_')` is all that is required to place ligatures in GDEF class 2.

Whereas here, when a glyph does not have those special anchors, you are classifying glyphs as ligatures based on the GlyphData.xml "Ligature" sub-category. Is that what Glyph does?